src/ApparentHorizons/Strahlkorper.cpp

```diff
@@ -0,0 +1,399 @@
+
+template <typename Fr>
+Strahlkorper<Fr>::Strahlkorper(const size_t l_max, const size_t m_max,
+                               const double radius,
+                               std::array<double, 3> center) noexcept
+    : l_max_(l_max),
+      m_max_(m_max),
+      ylm_(l_max, m_max),
+      // clang-tidy: do not std::move trivially constructable types
+      center_(std::move(center)),  // NOLINT
+      strahlkorper_coefs_(ylm_.spectral_size(), 0.0) {
+  ylm_.add_constant(&strahlkorper_coefs_, radius);
+}
+
+template <typename Fr>
+Strahlkorper<Fr>::Strahlkorper(
+    const size_t l_max, const size_t m_max,
+    const Strahlkorper& another_strahlkorper) noexcept
+    : l_max_(l_max),
+      m_max_(m_max),
+      ylm_(l_max, m_max),
+      center_(another_strahlkorper.center_),
+      strahlkorper_coefs_(another_strahlkorper.ylm_.prolong_or_restrict(
+          another_strahlkorper.strahlkorper_coefs_, ylm_)) {}
+
+template <typename Fr>
+Strahlkorper<Fr>::Strahlkorper(const Strahlkorper& another_strahlkorper,
+                               DataVector coefs) noexcept
+    : l_max_(another_strahlkorper.l_max_),
+      m_max_(another_strahlkorper.m_max_),
+      ylm_(another_strahlkorper.ylm_),
+      center_(another_strahlkorper.center_),
+      strahlkorper_coefs_(std::move(coefs)) {
+  ASSERT(
+      strahlkorper_coefs_.size() == another_strahlkorper.ylm_.spectral_size(),
+      "Bad size " << strahlkorper_coefs_.size() << ", expected "
+                  << another_strahlkorper.ylm_.spectral_size());
+}
+
+template <typename Fr>
+Strahlkorper<Fr>::Strahlkorper(const DataVector& radius_at_collocation_points,
+                               const size_t l_max, const size_t m_max,
+                               std::array<double, 3> center) noexcept
+    : l_max_(l_max),
+      m_max_(m_max),
+      ylm_(l_max, m_max),
+      // clang-tidy: do not std::move trivially constructable types
+      center_(center),  // NOLINT
+      strahlkorper_coefs_(ylm_.phys_to_spec(radius_at_collocation_points)) {
+  ASSERT(radius_at_collocation_points.size() == ylm_.physical_size(),
+         "Bad size " << radius_at_collocation_points.size() << ", expected "
+                     << ylm_.physical_size());
+}
+
+template <typename Fr>
+void Strahlkorper<Fr>::pup(PUP::er& p) {
+  p | l_max_;
+  p | m_max_;
+  p | center_;
+  p | strahlkorper_coefs_;
+  // All variables other than the above are cached quantities that
+  // can be recomputed from the above.
+
+  if (p.isUnpacking()) {
+    ylm_ = YlmSpherepack(l_max_, m_max_);
+  }
+}
+
+template <typename Fr>
+std::array<double, 3> Strahlkorper<Fr>::physical_center() const noexcept {
+  // For better accuracy, one should do an integral of
+  // x^i * AreaElement divided by Area.  The method here is an approximation
+  // using only the l=1 modes.
+  // See Eq. (37) of Hemberger et al, arXiv:1211.6079 for the form of the
+  // exact physical center that we approximate here.
```

Comment:
I would phrase it more similarly to the header file: eqns 38-40 are implemented for approx result, eqn 37 gives exact value.